### DISCUSSION TITLE
Signal / Whatsapp integration

### DISCUSSION THREAD

**Discussion by pmatuszy** [healthchecks/healthchecks#758] [General]
... and yest another question:
how to do the integration with Signal / Whatsapp? I can't see any section in the documentation explaining this...
Paul

**Comment by cuu508** [comment]
Signal: https://healthchecks.io/docs/self_hosted_configuration/#SIGNAL_CLI_SOCKET

WhatsApp: see TWILIO_* settings: https://healthchecks.io/docs/self_hosted_configuration/#TWILIO_ACCOUNT

For Signal, you will first want to get familiar with running signal-cli, and sending messages from command line. I would start with reading signal-cli [README](https://github.com/AsamK/signal-cli/blob/master/README.md), [man page](https://github.com/AsamK/signal-cli/blob/master/man/signal-cli.1.adoc) and [Wiki pages](https://github.com/AsamK/signal-cli/wiki).

Healthchecks uses Twilio to send WhatsApp messages. You specify Twilio API credentials and phone number in the TWILIO_ env variables and off you go. But there's a non-trivial onboarding process you have to go through in your Twilio account to enable WhatsApp message sending. I'm not sure if it's even possible if you're an individual, not a company.

**Reply by pmatuszy** [reply]
Thank you very much for your answer.

I am pretty familiar with signal-cli. I set it to start on the file socket. I've checked and it works there.
When I add integration I get HTTP 500 ERROR which in debug mode show this:



```Request URL:	https://192.168.1.249:7900/integrations/6cdfb8ed-758f-4e09-9992-70114534a100/test/
Django Version:	4.1.2
Exception Type:	TypeError
Exception Value:	
argument of type 'NoneType' is not iterable
Exception Location:	/encrypted/webapps/healthchecks/hc/api/transports.py, line 49, in get_nested
Raised during:	hc.front.views.send_test_notification
Python Executable:	/encrypted/webapps/hc-venv/bin/python
Python Version:	3.10.6
Python Path:	
['/encrypted/webapps/healthchecks',
 '/encrypted/webapps/hc-venv/bin',
 '/usr/lib/python310.zip',
 '/usr/lib/python3.10',
 '/usr/lib/python3.10/lib-dynload',
 '/encrypted/webapps/hc-venv/lib/python3.10/site-packages']
Server time:	Wed, 21 Dec 2022 08:35:02 +0000

```
integration is done with my cell phone number - I've double-checked it and it looks good
![image](https://user-images.githubusercontent.com/58336955/208860073-291ff10c-f063-4638-9ba6-a01308e9d977.png)

Have you got any idea why I can't get this integration working?

**Reply by cuu508** [reply]
Not 100% sure but I think what happens is signal-cli RPC returns a message in a format that Healthchecks does not expect, and Healthchecks throws an exception. 

I'd like to fix Healthchecks to handle it gracefully, and provide an useful error message to the user. I just created an issue – #763.

Would you be able to add a debug statement in the code, to print out the message that Healthchecks receives from signal-cli? In `hc/api/transports.py`, around line 904 add `print(reply)`:


before: 

```python
        for reply_bytes in self._read_replies(payload_bytes):
            try:
                reply = json.loads(reply_bytes.decode())
            except ValueError:
                raise TransportError("signal-cli call failed (unexpected response)")

            if reply.get("id") == payload["id"]:
                if "error" not in reply:
                    # success!
                    break
```

after:

```python
        for reply_bytes in self._read_replies(payload_bytes):
            try:
                reply = json.loads(reply_bytes.decode())
            except ValueError:
                raise TransportError("signal-cli call failed (unexpected response)")

            print(reply)  # <--- add this
            if reply.get("id") == payload["id"]:
                if "error" not in reply:
                    # success!
                    break
```

**Reply by pmatuszy** [reply]
ok - added.

![image](https://user-images.githubusercontent.com/58336955/208872975-a3f71590-f2a0-45e6-8c2e-aa455d7857ca.png)

it fails on line # 49
![image](https://user-images.githubusercontent.com/58336955/208873429-c5263053-7983-4261-aeda-6969e02d9093.png)
and my line 49 is below:
![image](https://user-images.githubusercontent.com/58336955/208873198-b18d426a-fda6-4aad-968b-098b22261c44.png)

will it help you?

**Reply by cuu508** [reply]
After adding the print statement, and triggering the error again, you should see the RPC message from signal-cli printed in the web server's stdout. Can you post it here?

**Comment by pmatuszy** [comment]
ok sorted. The problem was that I was starting signal-cli without "-a" parameter.
Lessons learned from the exercise:
1. how to start signal-cli?
```
signal-cli -a +4xxxaaabbbb daemon --socket /tmp/signal-socket
```
2. set ACL for that file so the user under which credentials WWW server starts has privs to that file
```
setfacl -m u:USERNAME:rw /tmp/signal-socket
```

having this in place integration works

Thank you once again for your great, quick and effective support!

**Reply by cuu508** [reply]
Awesome, happy to hear that. 

I [just fixed](https://github.com/healthchecks/healthchecks/commit/a49bc4ef3a9bf4f6e0557b41f00786438ec8dd19) the Signal integration to handle the missing "-a" scenario better. It will now return an error message that says: "Could not send a test notification. signal-cli call failed (-32602)." 

And, semi-related, I added the [ADMINS](https://github.com/healthchecks/healthchecks/commit/ae53aaaa3a9f3e420858f9b877dd5047fae689ee) environment variable. If you specify an email address in this variable, Healthchecks will send error reports to this address when DEBUG=False. And so, if you hit a HTTP 500 again, you will not have to turn DEBUG on just to see the error details.

**Reply by cuu508** [reply]
Started a discussion about making "-a" optional here: https://github.com/AsamK/signal-cli/discussions/1132
